### ISSUE TITLE
Potential Bug: Appearance head not running through the classifier guidance updates

### ISSUE THREAD

**Issue by cs-mshah** [google-research/readout_guidance#2]
I tried skimming through the codebase manually and with a debugger for the `script_drag.py` file. I noticed that the classifier guidance like updates are only made if the `controller.has_edits()`.

```python
if controller.has_edits(): 
        latents, losses = optimize_latents(latents)
        if losses:
            rg_loss = np.mean(losses)
        else:
            rg_loss = "N/A"
```
This is the method from the `ReadoutGuidance` class:

```python
def has_edits(self):
        return any(["control" in edit or "perceptual" in edit or "points" in edit for edit in self.edits])
```

If we see the `script_drag.py` file, in L149, `edits = set_edits_image(edits, first_frame)`, we have:
```python
def set_edits_image(edits, image):
    for edit in edits:
        if edit["head_type"] == "appearance":
            edit["control_image"] = image
    return edits
```

which basically means that the `"control"` is never set in the edits config dict and the latents are never optimised. Even after setting `edit["control"] = image`, I was unable to find where in the loss function is the control used as the ground truth for computing the readout loss.

```python
def loss_guidance(controller, feats, batch_idx, gt_idx, edits=[], log=False, emb=None, latents_scale=None, t=None, i=None):
    rg_loss = 0
    for edit in edits:
        aggregation_network = edit.get("aggregation_network", None)
        gt_feat = feats[gt_idx][None, ...].detach().clone()
        obs_feat = feats[batch_idx][None, ...]
        # Run aggregation network with the option to
        # deactive the timestep embedding conditioning (use_emb)
        obs_feat, gt_feat = run_aggregation(
            obs_feat, 
            gt_feat, 
            aggregation_network, 
            emb if edit.get("use_emb", True) else None
        )
        if edit["head_type"] == "spatial":
            spa_loss, obs_feat, gt_feat = loss_spatial(obs_feat, gt_feat, edit, latents_scale)
            rg_loss += spa_loss * get_loss_rescale(edit, i)
        elif edit["head_type"] == "appearance":
            app_loss, obs_feat, gt_feat = loss_appearance(obs_feat, gt_feat, edit, latents_scale)
            rg_loss += app_loss * get_loss_rescale(edit, i)
        elif edit["head_type"] == "correspondence":
            corr_loss, obs_feat, gt_feat = loss_correspondence(obs_feat, gt_feat, edit)
            rg_loss += corr_loss * get_loss_rescale(edit, i)
        if log:
            feat = feats_to_rgb(torch.concatenate([obs_feat, gt_feat], dim=0), latents_scale)
            if feat is not None:
                controller.obs_feat = Image.fromarray(feat[0])
                controller.gt_feat = Image.fromarray(feat[1])
            else:
                controller.obs_feat = None
                controller.gt_feat = None
    return rg_loss
```
```python
def loss_appearance(obs_feat, gt_feat, edit, latents_scale=(0, 1)):
    gt_feat = einops.rearrange(gt_feat, 'b w h c -> b c w h')
    obs_feat = einops.rearrange(obs_feat, 'b w h c -> b c w h')
    if edit.get("pool", False):
        obs_feat = obs_feat.mean(dim=[-2, -1])
    if edit.get("bbox") is not None:
        # Constraint the appearance similarity loss
        # within some pre-defined bounding box
        gt_bbox = edit["bbox"][0]
        obs_bbox = edit["bbox"][1]
        _, _, w, h = obs_feat.shape
        obs_feat = obs_feat[:, :, obs_bbox[1]:obs_bbox[3], obs_bbox[0]:obs_bbox[2]]
        gt_feat = gt_feat[:, :, gt_bbox[1]:gt_bbox[3], gt_bbox[0]:gt_bbox[2]]
        obs_feat = torch.nn.functional.interpolate(obs_feat, size=(w, h), mode="bilinear")
        gt_feat = torch.nn.functional.interpolate(gt_feat, size=(w, h), mode="bilinear")
    loss_spatial = 1 - torch.nn.functional.cosine_similarity(obs_feat, gt_feat, dim=1)
    loss = loss_spatial.mean()
    loss = loss * (latents_scale[1] - latents_scale[0]) ** 2
    return loss, obs_feat, gt_feat
```

the `feats` come from the decoder after passing the latents through the `unet`. How is the control being used for the case of appearance.

**Comment by g-luo** [comment]
Hi! Thanks for raising this issue; the code indeed does not propagate the name change from `perceptual` to `appearance`. I've updated the [code](https://github.com/google-research/readout_guidance/blob/bd994f1757a07285628edd6bf92be8d1373460d1/readout_guidance/rg_pipeline.py#L61) so that the name of the readout head does not matter for applying guidance updates.

**However, all code in the initial release are actually applying appearance guidance and producing the correct results.** In the pre-loaded [configs](https://github.com/google-research/readout_guidance/blob/main/configs/drag_real.yaml), the appearance head is always used in conjunction with the correspondence head,  so you'll notice that with this config the original function actually returns `True`. Note that the correspondence head raises the flag `"points"` and the pose/depth/edge heads raise the flag `"control"`. This mainly becomes an issue with new code that may use the appearance head alone.

**Comment by trwang477** [comment]
Could you also please address the question of how the control image is used as the ground truth for computing the readout loss? I'm trying to recreate sampling code for identity consistency, how could the conditioning image `edit["control"] = image` control the generation process in current code, as the loss is calculated w.r.t. latent at t not an image? Thank you in advance.
